src/messaging/ExchangeContext.h

```diff
@@ -0,0 +1,360 @@
+     *  @retval  #CHIP_NO_ERROR                             if the CHIP layer successfully sent the message down to the
+     *                                                       network layer.
+     */
+    CHIP_ERROR SendMessage(uint32_t protocolId, uint8_t msgType, System::PacketBuffer * msgPayload, uint16_t sendFlags = 0,
+                           void * msgCtxt = nullptr);
+
+    /**
+     *  Handle a received CHIP message on this exchange.
+     *
+     *  @param[in]    packetHeader  A reference to the PacketHeader object.
+     *
+     *  @param[in]    payloadHeader A reference to the PayloadHeader object.
+     *
+     *  @param[in]    msgBuf        A pointer to the PacketBuffer object holding the CHIP message.
+     *
+     *  @retval  #CHIP_ERROR_INVALID_ARGUMENT               if an invalid argument was passed to this HandleMessage API.
+     *  @retval  #CHIP_ERROR_INCORRECT_STATE                if the state of the exchange context is incorrect.
+     *  @retval  #CHIP_NO_ERROR                             if the CHIP layer successfully delivered the message up to the
+     *                                                       protocol layer.
+     */
+    CHIP_ERROR HandleMessage(const PacketHeader & packetHeader, const PayloadHeader & payloadHeader, System::PacketBuffer * msgBuf);
+
+    /**
+     *  Handle a received CHIP message on this exchange.
+     *
+     *  @param[in]    packetHeader  A reference to the PacketHeader object.
+     *
+     *  @param[in]    payloadHeader A reference to the PayloadHeader object.
+     *
+     *  @param[in]    msgBuf        A pointer to the PacketBuffer object holding the CHIP message.
+     *
+     *  @param[in]    umhandler     A unsolicited message callback handler.
+     *
+     *  @retval  #CHIP_ERROR_INVALID_ARGUMENT               if an invalid argument was passed to this HandleMessage API.
+     *  @retval  #CHIP_ERROR_INCORRECT_STATE                if the state of the exchange context is incorrect.
+     *  @retval  #CHIP_NO_ERROR                             if the CHIP layer successfully delivered the message up to the
+     *                                                       protocol layer.
+     */
+    CHIP_ERROR HandleMessage(const PacketHeader & packetHeader, const PayloadHeader & payloadHeader, System::PacketBuffer * msgBuf,
+                             ExchangeContext::MessageReceiveFunct umhandler);
+
+    /**
+     *  Search for an existing exchange that the message applies to.
+     *
+     *  @param[in]    packetHeader  A reference to the PacketHeader object.
+     *
+     *  @param[in]    payloadHeader A reference to the PayloadHeader object.
+     *
+     *  @retval  true                                       If a match is found.
+     *  @retval  false                                      If a match is not found.
+     */
+    bool MatchExchange(const PacketHeader & packetHeader, const PayloadHeader & payloadHeader);
```

Comment:
MatchExchange is defined in ExchangeContext, but called from ExchangeManager only.

Reply:
Let's make this private then, keep it out of headers

Reply:
If make this function private in ExchangeContext, how to call it from ExchangeManager? Unless declare ExchangeManger as the friend of ExchangeContext, but it is not recommended to use friend.

Reply:
Maybe the documentation is a bit odd - I would expect searches to return an iterator. This seems to be a 'are these headers belonging to this exchange context'. Is this true? The passed in information seems a bit large - includes all headers. Coud there be a subset of items to check for? I would imagine we only need 'node id' and 'exchange id'. Is that true? if so, could we pass those rather than the entire headers? 

Regarding friend usage: why is this a member of the exchage? Could the context provide public GetPeerNodeId() and GetExchangeId() and have this matches inside either a helper or part of th exchange manager (those are actually already provided ... is that sufficient?)?

Reply:
Suggestions for improving this:
  - doccomment should mention what matching is done (presumably nodeid and exchange id) rather than 'search' as no search is performed
  - This should likely use 'getters' and be a anon namespace method within the cpp file rather than a member method.

Reply:
We only need to know if a message belongs to this exchange, something like:
if (ec->GetExchangeMgr() != nullptr && ec->MatchExchange(packetHeader, payloadHeader))
{
   .....
}

We need more info to do the match other than node-id and exchange-id, such as if the message was sent by an initiator and the exchange context is a responder OR if the message was sent by a responder and the exchange context is an initiator. Maybe more criteria will added later, so I suggest to keep this API aligned with WML.